### ISSUE TITLE
PSNR以及SSIM指标有差距

### ISSUE THREAD

**Issue by kesssssa** [Yaozzz666/DSCF-SR#6]
作者团队你们好，你们提出的模型是一个非常轻量化且高效的模型！我有一个问题，我在使用模型参数为131k的模型（x4）进行推理Set5数据集时，平均PSNR是29.844455dB，但你们的论文中给出的PSNR是31.73dB，为什么差了大约2dB呢？除此之外，SSIM指标也是有差距的，我使用的是你们github中给出的test_demo.py修改之后的代码。
这是我在推理Set5数据集时的推理过程信息：
LogHandlers setup!
25-07-02 20:28:47.871 : 23_DSCF
25-07-02 20:28:48.595 : baby.png - PSNR: 31.99 dB; SSIM: 0.8644.
25-07-02 20:28:48.852 : bird.png - PSNR: 32.16 dB; SSIM: 0.9192.
25-07-02 20:28:49.049 : butterfly.png - PSNR: 26.37 dB; SSIM: 0.8925.
25-07-02 20:28:49.278 : head.png - PSNR: 29.86 dB; SSIM: 0.7193.
25-07-02 20:28:49.501 : woman.png - PSNR: 28.84 dB; SSIM: 0.8994.
25-07-02 20:28:49.501 : Max Memory        : 418.513 [M]
25-07-02 20:28:49.502 : Average runtime   : 3.421542 milliseconds
25-07-02 20:28:49.502 : Average PSNR      : 29.844455 dB
Unsupported operator aten::sub encountered 7 time(s)
Unsupported operator aten::mul encountered 7 time(s)
Unsupported operator aten::silu_ encountered 12 time(s)
Unsupported operator aten::sigmoid encountered 6 time(s)
Unsupported operator aten::add encountered 6 time(s)
Unsupported operator aten::pixel_shuffle encountered 1 time(s)
25-07-02 20:28:49.540 : #Activations       : 38.9284 [M]
25-07-02 20:28:49.540 : #Conv2d            : 22
25-07-02 20:28:49.541 : FLOPs              : 8.5350 [G]
25-07-02 20:28:49.541 : #Params            : 0.1308 [M]

下面是我自己修改好的测试代码：
import os.path
import logging
import torch
import argparse
import json
import glob

from pprint import pprint
from fvcore.nn import FlopCountAnalysis
from utils.model_summary import get_model_activation
from utils import utils_logger
from utils import utils_image as util

def select_model(args, device):
    model_id = args.model_id
    if model_id == 0:
        from models.team00_EFDN import EFDN
        name, data_range = f"{model_id:02}_EFDN_baseline", 1.0
        model_path = os.path.join('model_zoo', 'team00_EFDN.pth')
        model = EFDN()
        model.load_state_dict(torch.load(model_path), strict=True)
    elif model_id == 23:
        from models.team23_DSCF import DSCF
        name, data_range = f"{model_id:02}_DSCF", 1.0
        model_path = os.path.join('model_zoo', 'team23_DSCF.pth')
        model = DSCF(3, 3, feature_channels=26, upscale=4)
        state_dict = torch.load(model_path)
        model.load_state_dict(state_dict, strict=False)
    else:
        raise NotImplementedError(f"Model {model_id} is not implemented.")

    model.eval()
    tile = None
    for v in model.parameters():
        v.requires_grad = False
    model = model.to(device)
    return model, name, data_range, tile

def select_dataset(data_dir):
    lr_paths = sorted(glob.glob(os.path.join(data_dir, "LR", "*.png")))
    hr_paths = sorted(glob.glob(os.path.join(data_dir, "HR", "*.png")))
    if len(lr_paths) != len(hr_paths):
        raise ValueError("Mismatch in number of LR and HR images.")
    return list(zip(lr_paths, hr_paths))

def forward(img_lq, model, tile=None, tile_overlap=32, scale=4):
    if tile is None:
        return model(img_lq)
    else:
        b, c, h, w = img_lq.size()
        tile = min(tile, h, w)
        stride = tile - tile_overlap
        sf = scale
        h_idx_list = list(range(0, h - tile, stride)) + [h - tile]
        w_idx_list = list(range(0, w - tile, stride)) + [w - tile]
        E = torch.zeros(b, c, h * sf, w * sf).type_as(img_lq)
        W = torch.zeros_like(E)
        for h_idx in h_idx_list:
            for w_idx in w_idx_list:
                in_patch = img_lq[..., h_idx:h_idx + tile, w_idx:w_idx + tile]
                out_patch = model(in_patch)
                out_patch_mask = torch.ones_like(out_patch)
                E[..., h_idx * sf:(h_idx + tile) * sf, w_idx * sf:(w_idx + tile) * sf].add_(out_patch)
                W[..., h_idx * sf:(h_idx + tile) * sf, w_idx * sf:(w_idx + tile) * sf].add_(out_patch_mask)
        return E.div_(W)

def run(model, model_name, data_range, tile, logger, device, args):
    sf = 4
    border = sf
    results = dict(runtime=[], psnr=[])
    if args.ssim:
        results['ssim'] = []

    data_path = select_dataset(args.data_dir)
    save_path = os.path.join(args.save_dir, model_name)
    util.mkdir(save_path)

    start = torch.cuda.Event(enable_timing=True)
    end = torch.cuda.Event(enable_timing=True)

    for img_lr_path, img_hr_path in data_path:
        img_name, ext = os.path.splitext(os.path.basename(img_hr_path))
        img_lr = util.imread_uint(img_lr_path, n_channels=3)
        img_lr = util.uint2tensor4(img_lr, data_range).to(device)

        start.record()
        img_sr = forward(img_lr, model, tile)
        end.record()
        torch.cuda.synchronize()
        results['runtime'].append(start.elapsed_time(end))

        img_sr_np = util.tensor2uint(img_sr, data_range)
        util.imsave(img_sr_np, os.path.join(save_path, img_name + ext))  # 保存图像

        img_hr = util.imread_uint(img_hr_path, n_channels=3).squeeze()
        img_hr = util.modcrop(img_hr, sf)
        psnr = util.calculate_psnr(img_sr_np, img_hr, border=border)
        results['psnr'].append(psnr)

        if args.ssim:
            ssim = util.calculate_ssim(img_sr_np, img_hr, border=border)
            results['ssim'].append(ssim)
            logger.info(f"{img_name + ext} - PSNR: {psnr:.2f} dB; SSIM: {ssim:.4f}.")
        else:
            logger.info(f"{img_name + ext} - PSNR: {psnr:.2f} dB")

    results['memory'] = torch.cuda.max_memory_allocated(torch.cuda.current_device()) / 1024 ** 2
    results['ave_runtime'] = sum(results['runtime']) / len(results['runtime'])
    results['ave_psnr'] = sum(results['psnr']) / len(results['psnr'])
    if args.ssim:
        results['ave_ssim'] = sum(results['ssim']) / len(results['ssim'])

    logger.info(f"Max Memory        : {results['memory']:.3f} [M]")
    logger.info(f"Average runtime   : {results['ave_runtime']:.6f} milliseconds")
    logger.info(f"Average PSNR      : {results['ave_psnr']:.6f} dB")
    return results

def main(args):
    utils_logger.logger_info("NTIRE2025-EfficientSR", log_path="NTIRE2025-EfficientSR.log")
    logger = logging.getLogger("NTIRE2025-EfficientSR")
    torch.cuda.current_device()
    torch.cuda.empty_cache()
    torch.backends.cudnn.benchmark = False
    device = torch.device("cuda" if torch.cuda.is_available() else "cpu")

    json_dir = os.path.join(os.getcwd(), "results.json")
    if not os.path.exists(json_dir):
        results = dict()
    else:
        with open(json_dir, "r") as f:
            results = json.load(f)

    model, model_name, data_range, tile = select_model(args, device)
    logger.info(model_name)

    result = run(model, model_name, data_range, tile, logger, device, args)

    input_dim = (3, 256, 256)
    activations, num_conv = get_model_activation(model, input_dim)
    flops = FlopCountAnalysis(model, torch.rand(1, *input_dim).to(device)).total() / 1e9
    num_params = sum(p.numel() for p in model.parameters()) / 1e6

    logger.info(f"#Activations       : {activations / 1e6:.4f} [M]")
    logger.info(f"#Conv2d            : {num_conv}")
    logger.info(f"FLOPs              : {flops:.4f} [G]")
    logger.info(f"#Params            : {num_params:.4f} [M]")

    results[model_name] = result
    results[model_name].update({
        "activations": activations / 1e6,
        "num_conv": num_conv,
        "flops": flops,
        "num_parameters": num_params
    })

    with open(json_dir, "w") as f:
        json.dump(results, f)

if __name__ == "__main__":
    parser = argparse.ArgumentParser("NTIRE2025-EfficientSR")
    parser.add_argument("--data_dir", required=True, type=str, help="path to dataset directory, should contain LR/ and HR/")
    parser.add_argument("--save_dir", default="./results", type=str, help="directory to save output images")
    parser.add_argument("--model_id", default=23, type=int)
    parser.add_argument("--ssim", action="store_true", help="calculate SSIM")
    args = parser.parse_args()
    pprint(args)
    main(args)

# 运行指令
# python test_demo_auto.py --data_dir ./data/Set5 --save_dir ./result --model_id 23 --ssim

**Comment by Yaozzz666** [comment]
你好！抱歉过了这么久才回复。test_demo 是 NTIRE 官方的测试代码，所以 PSNR 和 SSIM 是在RGB 通道测试的，而论文中是在 Y 通道上测试的这两个指标，所以会有一定的差距。希望我的回答足够清楚。
